### ISSUE TITLE
if a clause throws something but specifies the wrong effect type then the error message is broken

### ISSUE THREAD

**Issue by kach** [accordproject/ergo#425]
This compiles fine:
```
namespace org.accordproject.payment.onetime

import org.accordproject.cicero.contract.*
import org.accordproject.cicero.runtime.*
import org.accordproject.money.MonetaryAmount

define transaction InitRequest extends Request {}
define transaction InitResponse extends Response {}

define transaction PaymentReceivedRequest extends Request {}
define transaction PaymentReceivedResponse extends Response {}

define enum ContractLifecycleStatus {
  UNINITIALIZED,
  OBLIGATION_EMITTED,
  COMPLETED
}

define asset OneTimePaymentState extends AccordContractState {
  status : ContractLifecycleStatus
}

define asset OneTimePaymentContract extends AccordContract {
  buyer : AccordParty,
  seller : AccordParty,
  initialAmount : MonetaryAmount
}

contract OneTimePayment over OneTimePaymentContract state OneTimePaymentState {
  clause init(request : InitRequest) : InitResponse emits PaymentObligation {
    emit PaymentObligation{
      contract: contract.contractId,
      promisor: contract.buyer,
      promisee: contract.seller,
      amount: contract.initialAmount,
      description: contract.buyer.partyId ++ " should pay contract amount to " ++ contract.seller.partyId
    };
    set state OneTimePaymentState {
      stateId: "1",
      status: "OBLIGATION_EMITTED"
    };
    return InitResponse{}
  }

  /**
   * Update the contract state once we've recieved a payment
   */
  clause paymentReceived(request : PaymentReceivedRequest): PaymentReceivedResponse
  throws ErgoErrorResponse
  {
    enforce (state.status = "OBLIGATION_EMITTED") ;
    set state OneTimePaymentState {
      stateId: state.stateId,
      status: "COMPLETED"
    };
    return PaymentReceivedResponse{}
  }
}
```
But this doesn't:
```
namespace org.accordproject.payment.onetime

import org.accordproject.cicero.contract.*
import org.accordproject.cicero.runtime.*
import org.accordproject.money.MonetaryAmount

define transaction InitRequest extends Request {}
define transaction InitResponse extends Response {}

define transaction PaymentReceivedRequest extends Request {}
define transaction PaymentReceivedResponse extends Response {}

define enum ContractLifecycleStatus {
  UNINITIALIZED,
  OBLIGATION_EMITTED,
  COMPLETED
}

define asset OneTimePaymentState extends AccordContractState {
  status : ContractLifecycleStatus
}

define asset OneTimePaymentContract extends AccordContract {
  buyer : AccordParty,
  seller : AccordParty,
  initialAmount : MonetaryAmount
}

contract OneTimePayment over OneTimePaymentContract state OneTimePaymentState {
  clause init(request : InitRequest) : InitResponse emits PaymentObligation {
    emit PaymentObligation{
      contract: contract.contractId,
      promisor: contract.buyer,
      promisee: contract.seller,
      amount: contract.initialAmount,
      description: contract.buyer.partyId ++ " should pay contract amount to " ++ contract.seller.partyId
    };
    set state OneTimePaymentState {
      stateId: "1",
      status: "OBLIGATION_EMITTED"
    };
    return InitResponse{}
  }

  /**
   * Update the contract state once we've recieved a payment
   */
  clause paymentReceived(request : PaymentReceivedRequest): PaymentReceivedResponse
  throws MonetaryAmount
  {
    enforce (state.status = "OBLIGATION_EMITTED") ;
    set state OneTimePaymentState {
      stateId: state.stateId,
      status: "COMPLETED"
    };
    return PaymentReceivedResponse{}
  }
}
```
and the error message is

> `TypeError (at line 48 col 2). This function is annotated to return {emit: Event[], response: PaymentReceivedResponse, state: OneTimePaymentState ..}? but actually returns {emit: Event[], response: PaymentReceivedResponse, state: OneTimePaymentState}?`

which is confusing because it doesn't mention the effect type at all.

The more sinister version of this is when you don't specify an effect type at all… I don't understand the behavior at all. It appears to depend on the existence of an `init()` clause? Very confusing.

**Comment by kach** [comment]
@jeromesimeon this is urgent-ish for the type system, and possibly related to the type inference change you pushed recently.

**Comment by jeromesimeon** [comment]
This looks to me like a mismatch between the declared error (in `throws`) and the default error for the `enforce`. This compiles fine:
```
namespace org.accordproject.payment.onetime

import org.accordproject.cicero.contract.*
import org.accordproject.cicero.runtime.*
import org.accordproject.money.MonetaryAmount

define transaction InitRequest extends Request {}
define transaction InitResponse extends Response {}

define transaction PaymentReceivedRequest extends Request {}
define transaction PaymentReceivedResponse extends Response {}

define enum ContractLifecycleStatus {
  UNINITIALIZED,
  OBLIGATION_EMITTED,
  COMPLETED
}

define asset OneTimePaymentState extends AccordContractState {
  status : ContractLifecycleStatus
}

define asset OneTimePaymentContract extends AccordContract {
  buyer : AccordParty,
  seller : AccordParty,
  initialAmount : MonetaryAmount
}

contract OneTimePayment over OneTimePaymentContract state OneTimePaymentState {
  clause init(request : InitRequest) : InitResponse emits PaymentObligation {
    emit PaymentObligation{
      contract: contract.contractId,
      promisor: contract.buyer,
      promisee: contract.seller,
      amount: contract.initialAmount,
      description: contract.buyer.partyId ++ " should pay contract amount to " ++ contract.seller.partyId
    };
    set state OneTimePaymentState {
      stateId: "1",
      status: "OBLIGATION_EMITTED"
    };
    return InitResponse{}
  }

  /**
   * Update the contract state once we've recieved a payment
   */
  clause paymentReceived(request : PaymentReceivedRequest): PaymentReceivedResponse
  throws MonetaryAmount
  {
    enforce (state.status = "OBLIGATION_EMITTED")
		else throw MonetaryAmount{
			doubleValue : 0.0,
			currencyCode : "USD"
		};
    set state OneTimePaymentState {
      stateId: state.stateId,
      status: "COMPLETED"
    };
    return PaymentReceivedResponse{}
  }
}
```

So I think the behavior is consistent, but clearly this is confusing.

There are a number of possible fixes, none of which is a clear winner to me:
1. synthesize an error value of the right type for enforce
2. remove `throws` from the clause signature and always return errors of the same type -- would work best if `throws` is revised to only construct values of the right type (maybe just with a message in it).
3. force the presence of an `else` clause in `enforce`

Any other suggestion?

I'm learning toward 2. which simplifies error handling while still giving the ability to pass a specific message as part of the enforce call or for other kinds of errors.

**Comment by kach** [comment]
The proper fix is to correctly infer the effect type the way we correctly infer response type, and give a better error message if there happens to be a declared effect type that doesn't match the inferred type.

Forcing a single error type with only a string message is not good; people want to put other info in an error for debugging purposes.

**Comment by jeromesimeon** [comment]
Fixed in #430
